plugins/SlicerT/SlicerT.cpp

```diff
@@ -0,0 +1,402 @@
+Plugin::Descriptor PLUGIN_EXPORT slicert_plugin_descriptor = {
+	LMMS_STRINGIFY(PLUGIN_NAME),
+	"SlicerT",
+	QT_TRANSLATE_NOOP("PluginBrowser", "Basic Slicer"),
+	"Daniel Kauss Serna <daniel.kauss.serna@gmail.com>",
+	0x0100,
+	Plugin::Type::Instrument,
+	new PluginPixmapLoader("icon"),
+	nullptr,
+	nullptr,
+};
+} // end extern
+
+// ################################# SlicerT ####################################
+
+SlicerT::SlicerT(InstrumentTrack* instrumentTrack)
+	: Instrument(instrumentTrack, &slicert_plugin_descriptor)
+	, m_noteThreshold(0.6f, 0.0f, 2.0f, 0.01f, this, tr("Note threshold"))
+	, m_fadeOutFrames(400.0f, 0.0f, 8192.0f, 1.0f, this, tr("FadeOut"))
+	, m_originalBPM(1, 1, 999, this, tr("Original bpm"))
+	, m_sliceSnap(this, tr("Slice snap"))
+	, m_enableSync(false, this, tr("BPM sync"))
+	, m_originalSample()
+	, m_parentTrack(instrumentTrack)
+{
+	m_sliceSnap.addItem("Off");
+	m_sliceSnap.addItem("1/1");
+	m_sliceSnap.addItem("1/2");
+	m_sliceSnap.addItem("1/4");
+	m_sliceSnap.addItem("1/8");
+	m_sliceSnap.addItem("1/16");
+	m_sliceSnap.addItem("1/32");
+	m_sliceSnap.setValue(0);
+}
+
+void SlicerT::playNote(NotePlayHandle* handle, sampleFrame* workingBuffer)
+{
+	if (m_originalSample.frames() <= 1) { return; }
+
+	const int noteIndex = handle->key() - m_parentTrack->baseNote();
+	const fpp_t frames = handle->framesLeftForCurrentPeriod();
+	const f_cnt_t offset = handle->noteOffset();
+	const int bpm = Engine::getSong()->getTempo();
+	const float pitchRatio = 1 / pow(2, m_parentTrack->pitchModel()->value() / 1200);
+
+	float speedRatio = static_cast<float>(m_originalBPM.value()) / bpm;
+	if (!m_enableSync.value()) { speedRatio = 1; }
+	speedRatio *= pitchRatio;
+	speedRatio *= Engine::audioEngine()->processingSampleRate() / static_cast<float>(m_originalSample.sampleRate());
+
+	float sliceStart, sliceEnd;
+	if (noteIndex > m_slicePoints.size() - 2 || noteIndex < 0) // full sample if ouside range
+	{
+		sliceStart = 0;
+		sliceEnd = 1;
+	}
+	else
+	{
+		sliceStart = m_slicePoints[noteIndex];
+		sliceEnd = m_slicePoints[noteIndex + 1];
+	}
+
+	if (!handle->m_pluginData) { handle->m_pluginData = new PlaybackState(sliceStart); }
+
+	float noteDone = ((PlaybackState*)handle->m_pluginData)->noteDone();
+	float noteLeft = sliceEnd - noteDone;
+
+	if (noteLeft > 0)
+	{
+		int noteFrame = noteDone * m_originalSample.frames();
+
+		SRC_STATE* resampleState = ((PlaybackState*)handle->m_pluginData)->resamplingState();
+		SRC_DATA resampleData;
+		resampleData.data_in = (float*)(m_originalSample.data() + noteFrame);
+		resampleData.data_out = (float*)(workingBuffer + offset);
+		resampleData.input_frames = noteLeft * m_originalSample.frames();
+		resampleData.output_frames = frames;
+		resampleData.src_ratio = speedRatio;
+
+		src_process(resampleState, &resampleData);
+
+		float nextNoteDone = noteDone + frames * (1.0f / speedRatio) / m_originalSample.frames();
+		((PlaybackState*)handle->m_pluginData)->setNoteDone(nextNoteDone);
```

Comment:
- 
```suggestion
		((static_cast<PlaybackState*>)handle->m_pluginData)->setNoteDone(nextNoteDone);
```

Reply:
Did you mean
```c++
static_cast<PlaybackState*>(handle->m_pluginData)->setNoteDone(nextNoteDone);
```

Reply:
Yes, sorry, I'm on mobile and at work 🫠.

Reply:
Done